projects/igniteui-angular/migrations/common/setup.spec.ts

```diff
@@ -0,0 +1,82 @@
+import { readFileSync } from 'fs';
+import { join } from 'path';
+
+import { EmptyTree } from '@angular-devkit/schematics';
+import { UnitTestTree } from '@angular-devkit/schematics/testing';
+import * as tss from 'typescript/lib/tsserverlibrary';
+import { serviceContainer } from './service-container';
+
+const configJson = {
+    version: 1,
+    projects: {
+        testProj: {
+            projectType: 'application',
+            root: '',
+            sourceRoot: 'testSrc',
+            architect: {
+                build: {
+                    builder: '@angular-devkit/build-angular:application',
+                    options: {
+                        browser: 'testSrc/appPrefix/component/test.component.ts'
+                    }
+                }
+            }
+        }
+    },
+    schematics: {
+        '@schematics/angular:component': {
+            prefix: 'appPrefix'
+        }
+    }
+};
+
+const tsConfig = readFileSync('tsconfig.json');
+
+/**
+ * Internal extension to ensure newly created .ts files are added to the project service
+ * or their source cache is updated/invalidated if the name is reused in multiple tests
+ */
+class IgxUnitTestTree extends UnitTestTree {
+    public override create(path: string, content: string | Buffer): void {
+        super.create(path, content);
+        if (!path.endsWith('.ts') && !path.endsWith('.html')) return;
+
+        const configured = serviceContainer.configured && serviceContainer.projectService.configuredProjects.size;
+        if (configured) {
+            // rush host update
+            serviceContainer.serverHost.host = this;
+
+            const entryPath = tss.server.toNormalizedPath(join(process.cwd(), path))
+            const scriptInfo = serviceContainer.projectService?.getOrCreateScriptInfoForNormalizedPath(entryPath, false);
+            if (!scriptInfo) {
+                return;
+            }
+            // integrate incoming virtual file in project/program for LS to work:
+            const project = serviceContainer.projectService.configuredProjects.values().next().value as tss.server.ConfiguredProject;
+            if (!project.containsScriptInfo(scriptInfo)) {
+                // add first to prevent open from creating another project that'll be inferred
+                project.addMissingFileRoot(scriptInfo.fileName);
```

Comment:
Funnily enough, seems to work just fine with just the attach since the UpdateChanges will normally call `addMissingFileRoot` anyway. And this possibly better reflects the behavior when the file was actually present when creating the project.